src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/CollectionsMarshal.cs

```diff
@@ -16,5 +19,36 @@ public static class CollectionsMarshal
         /// </summary>
         public static Span<T> AsSpan<T>(List<T>? list)
             => list is null ? default : new Span<T>(list._items, 0, list._size);
+
+        public static ref TValue GetValueRef<TKey, TValue>(Dictionary<TKey, TValue> dictionary, [NotNull] TKey key) where TKey : notnull
+        {
+            ref TValue value = ref dictionary.FindValue(key);
+            if (Unsafe.IsNullRef(ref value))
+            {
+                ThrowHelper.ThrowKeyNotFoundException(key);
+            }
+
+            return ref value;
+        }
+
+        public static ref TValue GetValueRefOrNullRef<TKey, TValue>(Dictionary<TKey, TValue> dictionary, TKey key) where TKey : notnull
+            => ref dictionary.FindValue(key);
+
+        public static ref TValue GetValueRefOrAddDefault<TKey, TValue>(Dictionary<TKey, TValue> dictionary, [NotNull] TKey key, out bool exists) where TKey : notnull
+        {
+            ref TValue value = ref dictionary.FindValue(key);
+            if (Unsafe.IsNullRef(ref value))
+            {
+                exists = false;
+                dictionary[key] = default!;
+                value = ref dictionary.FindValue(key);
+            }
+            else
+            {
+                exists = true;
+            }
+
+            return ref value!;
+        }
```

Comment:
If this is the implementation, is it really worth adding? i.e. is this any better than a consumer could write themselves?

Reply:
Well... it could be better but that would need changes to `Dictionary.TryInsert` which will then need measurements it doesn't regress regular `Add`, `TryAdd` etc. Might just remove it for now 😉